src/planning.py

```diff
@@ -258,11 +276,51 @@ def _skeleton_generator(
         else:
             # Generate successors.
             metrics["num_nodes_expanded"] += 1
+            # If an abstract ldl policy is provided, generate policy-based
+            # successors first
+            if abstract_policy is not None:
+                current_node = node
+                for _ in range(sesame_max_policy_guided_rollout):
+                    if task.goal.issubset(current_node.atoms):
+                        break
+                    action = abstract_policy(current_node.atoms,
+                                             current_objects, task.goal)
+                    if action is None:
+                        break
+                    child_atoms = utils.apply_operator(action,
+                                                       set(current_node.atoms))
+                    child_skeleton = current_node.skeleton + [action]
+                    child_skeleton_tup = tuple(child_skeleton)
+                    if child_skeleton_tup in visited_skeletons:
+                        continue
+                    visited_skeletons.add(child_skeleton_tup)
+                    child_node = _Node(
+                        atoms=child_atoms,
+                        skeleton=child_skeleton,
+                        atoms_sequence=current_node.atoms_sequence +
+                        [child_atoms],
+                        parent=current_node)
+
+                    metrics["num_nodes_created"] += 1
+                    # priority is g [plan length] plus h [heuristic]
+                    priority = (len(child_node.skeleton) +
+                                heuristic(child_node.atoms))
+                    hq.heappush(queue,
+                                (priority, rng_prio.uniform(), child_node))
+                    current_node = child_node
+                    if time.time() - start_time >= timeout:
+                        break
+            # Generate primitive successors
             for nsrt in utils.get_applicable_operators(ground_nsrts,
                                                        node.atoms):
                 child_atoms = utils.apply_operator(nsrt, set(node.atoms))
+                child_skeleton = node.skeleton + [nsrt]
+                child_skeleton_tup = tuple(child_skeleton)
+                if child_skeleton_tup in visited_skeletons:
```

Comment:
This is just for when you are generating policy-based skeletons. A* won't generate duplicate skeletons when doing bilevel planning but it is possible for the policy to generate skeletons identical to the ones generated in bilevel planning. So, those lines of code necessary only if we're using a policy. Otherwise, they're redundant. I can just put them inside an if, that was a really good catch. Thank you!